system/datatypes/mailinglist.php

```diff
@@ -0,0 +1,959 @@
+		return $this->xml->documentElement;
+	}
+
+	public function build() {
+		$this->html->writeToElement('pagename', showPagename($this->pagename).' '.asterisk($this->privateFlag));
+
+		$firstArgument = $this->getArg(0);
+		if ('edit' != $firstArgument && !$this->hasSender()) {
+			notify('warning', __('ml-no-sender'));
+		}
+
+		switch ($firstArgument) {
+			case null:
+				return $this->indexAction();
+			case 'edit':
+				return $this->editAction();
+			case 'confirm':
+				return $this->confirmAction();
+			case 'unsubscribe':
+				return $this->unsubscribeAction();
+			case 'addresses':
+				return $this->addressesAction();
+			case 'create':
+				return $this->createMailingAction();
+			default:
+				switch ($this->getArg(1)) {
+					case 'edit':
+						return $this->editMailingAction($firstArgument);
+					default:
+						return $this->showMailingAction($firstArgument);
+				}
+		}
+	}
+
+	public function indexAction() {
+		// check if form is posted and get form data
+		$formData = [];
+		$formPosted = $this->isPosted(self::FORM_CMD_LIST_SUBSCRIBE);
+		if ($formPosted) {
+			$formData = $_POST;
+		}
+
+		// create form
+		$form = $this->createSubscribeForm($formData);
+
+		// process form if it was posted
+		if ($formPosted) {
+			$form->validateRequiredField(self::FIELD_NAME_EMAIL);
+			$form->validateEmailField(self::FIELD_NAME_EMAIL);
+			if (empty($form->errors)) {
+				// check if email is already in the list, no need to have it twice.
+				$this->xml->lockAndReload();
+				$email = strtolower($form->dataFor(self::FIELD_NAME_EMAIL));
+				/** @var HyphaDomElement $addresses */
+				$addresses = $this->getDoc()->getOrCreate('addresses');
+				/** @var NodeList $addressCollection */
+				$addressCollection = $addresses->children()->findXPath('//address[@email="' . $email . '"]');
+				$address = $addressCollection->first();
+
+				if (!$address instanceof HyphaDomElement) {
+					$addAddress = true;
+					$sendEmail = true;
+					$code = $this->constructCode();
+				} else {
+					// if the email is there just send another email if it is still pending
+					$sendEmail = $address->getAttribute('status') == self::ADDRESS_STATUS_PENDING;
+					$addAddress = false;
+					$code = $address->getAttribute('code');
+				}
+
+				if ($addAddress) {
+					/** @var HyphaDomElement $address */
+					// add email with status pending
+					$address = $this->xml->createElement('address');
+					$address->setAttribute('email', $email);
+					$address->setAttribute('status', self::ADDRESS_STATUS_PENDING);
+					$address->setAttribute('confirm-code', $code);
+					/** @var HyphaDomElement $addresses */
+					$addresses = $this->getDoc()->getOrCreate('addresses');
+					$addresses->append($address);
+					$this->xml->saveAndUnlock();
+				} else {
+					$this->xml->unlock();
+				}
+
+				if ($sendEmail) {
+					// send email so that action can be confirmed
+					$link = $this->constructFullPath(sprintf('%s/confirm?code=%s', $this->pagename, $code));
```

Comment:
pagename and code need to be passed through `urlencode` here, since it's hard to guarantee they  contain no breaking special characters (e.g. &).